Improve diagnostics for functions in `struct` definitions

LeSeulArtichaut opened the pull request:
Tries to implement #76421.
This is probably going to need unit tests, but I wanted to hear from review all the cases tests should cover.

I'd like to follow up with the "mechanically applicable suggestion here that adds an impl block" step, but I'd need guidance. My idea for now would be to try to parse a function, and if that succeeds, create a dummy `ast::Item` impl block to then format it using `pprust`. Would that be a viable approach? Is there a better alternative?

r? @matklad cc @estebank﻿

jyn514 commented:
It would be helpful to add a UI test for the example in https://github.com/rust-lang/rust/issues/76421 - even if it doesn't hit all the cases, it helps show that the diagnostic looks roughly right.

LeSeulArtichaut commented:
@jyn514 pushed the test

Dylan-DPC commented:
@matklad any updates?

camelid commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -1363,6 +1363,29 @@ impl<'a> Parser<'a> {
+                    lo.to(self.prev_token.span),
+                    "functions are not allowed in struct definitions",
+                );
+                err.help("unlike in C++, Java, and C#, functions are declared in `impl` blocks");
```
I think we should say:

```suggestion
                err.help("unlike in C++, Java, and C#, methods are declared in `impl` blocks");
```

instead because you can declare functions outside of `impl` blocks. Specifically, you declare *methods* in `impl` blocks.

estebank replied:
You can also have associated `fn`s with no receiver, equivalent to Java's static methods, which we don't call methods :)

camelid replied:
True, but we're saying that methods are declared in `impl` blocks, not that all functions in `impl` blocks are methods ;)

LeSeulArtichaut replied:
Should I write "methods and associated functions", or would that be confusing to a newcomer? Would it be ok to just write "methods"?

camelid replied:
I think just "methods" is fine since that's probably what the user intended.

jackh726 replied:
I think `functions` here are fine.

Dylan-DPC commented:
r? @estebank

bors commented:
:umbrella: The latest upstream changes (presumably #83188) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

crlf0710 commented:
@LeSeulArtichaut Triage: There's merge conflict.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2255100449) [(plain)](https://github.com/rust-lang/rust/commit/565e83a2362be20575061c174ff16245fa986b05/checks/2255100449/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking chalk-engine v0.55.0
error[E0061]: this function takes 1 argument but 0 arguments were supplied
    --> compiler/rustc_parse/src/parser/item.rs:1420:31
     |
1420 |             let err = if self.check_fn_front_matter() {
     |                               |
     |                               expected 1 argument
     |
note: associated function defined here
note: associated function defined here
    --> compiler/rustc_parse/src/parser/item.rs:1664:19
     |
1664 |     pub(super) fn check_fn_front_matter(&mut self, check_pub: bool) -> bool {

    Checking rustc_middle v0.0.0 (/checkout/compiler/rustc_middle)
error: aborting due to previous error


For more information about this error, try `rustc --explain E0061`.
error: could not compile `rustc_parse`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_codegen_ssa" "-p" "rustc_apfloat" "-p" "rustc_span" "-p" "rustc_arena" "-p" "rustc_index" "-p" "rustc_serialize" "-p" "rustc_hir" "-p" "rustc_feature" "-p" "rustc_symbol_mangling" "-p" "rustc_target" "-p" "rustc_ast" "-p" "rustc_lexer" "-p" "rustc_session" "-p" "rustc_lint_defs" "-p" "rustc_incremental" "-p" "rustc_graphviz" "-p" "rustc_fs_util" "-p" "rustc_errors" "-p" "rustc_macros" "-p" "rustc_data_structures" "-p" "rustc_middle" "-p" "rustc_query_system" "-p" "rustc_type_ir" "-p" "rustc_attr" "-p" "rustc_ast_pretty" "-p" "rustc_driver" "-p" "rustc_parse" "-p" "rustc_hir_pretty" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_ast_passes" "-p" "rustc_mir" "-p" "rustc_infer" "-p" "coverage_test_macros" "-p" "rustc_trait_selection" "-p" "rustc_parse_format" "-p" "rustc_save_analysis" "-p" "rustc_typeck" "-p" "rustc_error_codes" "-p" "rustc_mir_build" "-p" "rustc_interface" "-p" "rustc_privacy" "-p" "rustc_ast_lowering" "-p" "rustc_builtin_macros" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_passes" "-p" "rustc_resolve" "-p" "rustc_ty_utils" "-p" "rustc_query_impl" "-p" "rustc_traits" "-p" "rustc_plugin_impl" "-p" "rustc_lint" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:36
```

</details>

LeSeulArtichaut commented:
Rebased.

jackh726 commented on src/test/ui/structs/struct-fn-in-definition.rs:
```diff
@@ -0,0 +1,11 @@
+// Suggested in issue #76421
```
Maybe add a short comment to explain what we are trying to suggest here?

jackh726 reviewed (approved):
@LeSeulArtichaut Do you want to just try to get the the current suggestion merged and then make a followup PR with an even better error? If so, then r=me on this, with or without an additional comment on the test.

In future PRs, I would like to see tests for unions or enums too. Or feel free to add that to this PR if it works.

jackh726 commented:
r? @jackh726

jackh726 commented:
@bors delegate+

bors commented:
:v: @LeSeulArtichaut can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"jackh726","delegate":"LeSeulArtichaut"} -->

jackh726 commented:
@bors r+

bors commented:
:pushpin: Commit 6717f81b96aca75f5b811104ae75620274dad35d has been approved by `jackh726`

<!-- @bors r=jackh726 6717f81b96aca75f5b811104ae75620274dad35d -->
<!-- homu: {"type":"Approved","sha":"6717f81b96aca75f5b811104ae75620274dad35d","approver":"jackh726"} -->

LeSeulArtichaut commented:
@jackh726 Enums did not work, I added a new commit to this PR. I also added a better test description as you requested.

> Do you want to just try to get the the current suggestion merged and then make a followup PR with an even better error?

What would an "even better error" look like? Do you mean a machine-applicable suggestion that extracts the function to an `impl` block, as estebank suggested? If so I'd prefer that to be in a follow-up PR.

PS: you're too fast

jackh726 commented:
> What would an "even better error" look like? Do you mean a machine-applicable suggestion that extracts the function to an `impl` block, as estebank suggested? If so I'd prefer that to be in a follow-up PR.

Yeah, the machine-applicable suggestion.

> PS: _you're too fast_
Only sometimes :)

jackh726 commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -1124,11 +1124,11 @@ impl<'a> Parser<'a> {
                 let struct_def = if this.check(&token::OpenDelim(token::Brace)) {
                     // Parse a struct variant.
-                    let (fields, recovered) = this.parse_record_struct_body()?;
+                    let (fields, recovered) = this.parse_record_struct_body("struct")?;
```
Hmm, I see. If it comes up, a fairly easy fix might be to add a `struct` flag that is true in an actual struct and false in a enum variant.
